onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -1024,17 +1024,65 @@ Status TransposeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, co
 class ReshapeOpBuilder : public BaseOpBuilder {
  public:
   void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) override;
+  static Status AddReshapeOperator(ModelBuilder& model_builder, const Node& node,
+                                   const std::string& input, const std::vector<int32_t>& shape) ORT_MUST_USE_RESULT;
 
  private:
   bool IsOpSupportedImpl(ModelBuilder& model_builder, const Node& node) override;
-
   Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node) override ORT_MUST_USE_RESULT;
+  static bool CanSkipReshape(const Node& node);
 };
 
 void ReshapeOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) {
   model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());
 }
 
+// We can skip the Reshape if only node(s) in this graph using the output is Matmul or Gemm,
+// which will automatically flatten the shape
+/* static */ bool ReshapeOpBuilder::CanSkipReshape(const Node& node) {
```

Comment:
Not quite following this. Doesn't the MatMul or Gemm use the shape info to calculate the output shape? If so, how does it work if the Reshape has not run?

Our CPU implementation does this
```
Status MatMul<T>::Compute(OpKernelContext* ctx) const {
  concurrency::ThreadPool* thread_pool = ctx->GetOperatorThreadPool();

  const auto* a = ctx->Input<Tensor>(0);
  const auto* b = ctx->Input<Tensor>(1);

  MatMulComputeHelper helper;
  ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b->Shape()));
  ...
```

Reply:
Update the comments,